perf(formatter): use `ArenaVec` and `ArenaBox`

Dunqing opened the pull request:
This PR aims to replace the Formatter's standard `Vec` and `Box` usages with `ArenaVec` and `ArenaBox` to gain performance, which could reduce the `drop` overhead.

There are three key changes:

1. Change the Document's element from `Vec<FormatElement<'a>>`
```diff
pub struct Document<'a> {
-    elements: Vec<FormatElement<'a>>,
+    elements: &'a [FormatElement<'a>],
}
```

2. Change the `FormatElement::interned`'s argument from `Rc<[FormatElement<'a>]>` to `&'a [FormatElement<'a>]`
```diff
- pub struct Interned<'a>(Rc<[FormatElement<'a>]>);
+ pub struct Interned<'a>(&'a [FormatElement<'a>]);
```

3. Change the `FormatElement::variants`'s variants from `Box<[Box<[FormatElement<'a>]>]>` to `&'a [&'a [FormatElement<'a>]]`
```diff
pub struct BestFittingElement<'a> {
-    variants: Box<[Box<[FormatElement<'a>]>]>,
+    variants: &'a [&'a [FormatElement<'a>]],
}
```

All other changes stem from those mentioned above.

Replacing the `Rc` with `&a` reference approach was suggested and guided by @overlookmotel. Thank you!

leaysgur commented:
### Merge activity

* **Nov 18, 2:40 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing commented on crates/oxc_formatter/src/formatter/builders.rs:
```diff
@@ -2549,9 +2550,12 @@ impl<'ast> Format<'ast> for BestFitting<'_, 'ast> {
         }
 
+        let formatted_variants =
+            ArenaVec::from_iter_in(formatted_variants, f.context().allocator());
```
Unfortunately, this doesn't work because the above `VecBuffer` already borrows a mutable reference `f` and creates an ArenaVec that requires borrowing an immutable reference `f`, which violates the borrow rules

Dunqing committed:
perf(formatter): use `ArenaVec` and `ArenaBox` (#15420)

This PR aims to replace the Formatter's standard `Vec` and `Box` usages with `ArenaVec` and `ArenaBox` to gain performance, which could reduce the `drop` overhead.

There are three key changes:

1. Change the Document's element from `Vec<FormatElement<'a>>`
```diff
pub struct Document<'a> {
-    elements: Vec<FormatElement<'a>>,
+    elements: &'a [FormatElement<'a>],
}
```

2. Change the `FormatElement::interned`'s argument from `Rc<[FormatElement<'a>]>` to `&'a [FormatElement<'a>]`
```diff
- pub struct Interned<'a>(Rc<[FormatElement<'a>]>);
+ pub struct Interned<'a>(&'a [FormatElement<'a>]);
```

3. Change the `FormatElement::variants`'s variants from `Box<[Box<[FormatElement<'a>]>]>` to `&'a [&'a [FormatElement<'a>]]`
```diff
pub struct BestFittingElement<'a> {
-    variants: Box<[Box<[FormatElement<'a>]>]>,
+    variants: &'a [&'a [FormatElement<'a>]],
}
```

All other changes stem from those mentioned above.

Replacing the `Rc` with `&a` reference approach was suggested and guided by @overlookmotel. Thank you!
